File: ass1/q3/q3.py
import numpy as np
import pandas as pd
from scipy.stats import bartlett, levene, f_oneway, ttest_ind
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.stats.multicomp import pairwise_tukeyhsd, MultiComparison
from statsmodels.stats.multicomp import MultiComparison
from statsmodels.stats.weightstats import DescrStatsW

# Input data
data = {
    'coating': ['C1']*4 + ['C2']*4 + ['C3']*4 + ['C4']*4 + ['C5']*4,
    'conductivity': [
        143, 141, 150, 146, # C1
        152, 149, 137, 143, # C2
        134, 133, 132, 127, # C3
        129, 127, 132, 129, # C4
        147, 148, 144, 142, # C5
    ]
}
df = pd.DataFrame(data)

# Bartlett's test
print("\n=== Bartlett's Test ===")
coatings = [df[df.coating == br]['conductivity'] for br in df.coating.unique()]
bart_result = bartlett(*coatings)
print(f"Statistic: {bart_result.statistic:.3f}, p-value: {bart_result.pvalue:.3f}")

# Levene's test
print("\n=== Levene's Test ===")
lev_result = levene(*coatings)
print(f"Statistic: {lev_result.statistic:.3f}, p-value: {lev_result.pvalue:.3f}")

# Overall mean & treatment effects
print("\n=== Overall Mean ===")
overall_mean = df['conductivity'].mean()
print(f"Overall mean: {overall_mean:.2f}")

print("\n=== coating Means & Treatment Effects ===")
coating_means = df.groupby('coating')['conductivity'].mean().reset_index()
coating_means['treatment_effect'] = coating_means['conductivity'] - overall_mean
print(coating_means)

# One-way ANOVA
print("\n=== ANOVA Results ===")
anova = f_oneway(*coatings)
print(f"F-statistic: {anova.statistic:.3f}, p-value: {anova.pvalue:.3f}")
# Also, show statsmodels enhanced table:
model = smf.ols('conductivity ~ coating', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
print(anova_table)

# Multiple Comparisons: Tukey's HSD
print("\n=== Tukey's HSD ===")
multicomp = MultiComparison(df['conductivity'], df['coating'])
tukey_result = multicomp.tukeyhsd()
print(tukey_result.summary())

# Fisher's LSD (pairwise t-tests, unadjusted)
print("\n=== Fisher's LSD ===")
from itertools import combinations
for b1, b2 in combinations(df.coating.unique(), 2):
    stat, p = ttest_ind(df[df.coating == b1]['conductivity'], df[df.coating == b2]['conductivity'])
    print(f"{b1} vs {b2} p-value: {p:.3f}")

# Simultaneous Confidence Intervals (Tukey)
print("\n=== Tukey Confidence Intervals ===")
for res in tukey_result._results_table.data[1:]:
    print(f"Diff: {res[2]:.2f} CI: ({res[4]:.2f}, {res[5]:.2f})")

File: ass1/q3/q3.r
# Tegan Jain
# 23CS30065
# Assignment 1
# Question 3

# Install packages
if (!require(car)) install.packages("car", dependencies = TRUE)
library(car)
if (!require(DescTools)) install.packages("DescTools", dependencies = TRUE)
library(DescTools)
if (!require(agricolae)) install.packages("agricolae", dependencies = TRUE)
library(agricolae)

# Input the data
coating <- rep(c("C1", "C2", "C3", "C4", "C5"), each = 4)
conductivity <- c(
  143, 141, 150, 146, # C1
  152, 149, 137, 143, # C2
  134, 133, 132, 127, # C3
  129, 127, 132, 129, # C4
  147, 148, 144, 142 # C5
)
battery_data <- data.frame(coating, conductivity)

# Bartlett's Test
cat("\n=== Bartlett's Test (equal variance assumption) ===\n")
print(bartlett.test(conductivity ~ coating, data = battery_data))

# Levene's Test
cat("\n=== Levene's Test ===\n")
print(leveneTest(conductivity ~ coating, data = battery_data))

# Overall Mean & Treatment Effects
overall_mean <- mean(battery_data$conductivity)
coating_means <- aggregate(
  conductivity ~ coating,
  data = battery_data,
  FUN = mean
)
coating_means$effect <- coating_means$conductivity - overall_mean
cat("\n=== Overall Mean ===\n")
print(overall_mean)
cat("\n=== coating Means & Treatment Effects ===\n")
print(coating_means)

# One-way ANOVA
battery_aov <- aov(conductivity ~ coating, data = battery_data)
cat("\n=== ANOVA Results ===\n")
print(summary(battery_aov))

# Duncan's Multiple Range Test
cat("\n=== Duncan's Multiple Range Test ===\n")
print(duncan.test(battery_aov, "coating"))
